File: GraphCodeBert/examples/trial_mixed.py
# trial_mixed.py

import socket
import threading

# --------- Defective functions ---------
# Resource leak
def fetch_api(endpoint):
    sock = socket.socket()
    sock.connect(('api.com', 80))
    sock.send(endpoint.encode())
    # Missing sock.close()
    return sock.recv(1024)

# Null pointer dereference
def process_data(data):
    if data is None:
        return data.upper()  # Will throw AttributeError

# Concurrency issue
shared_counter = 0
def increment():
    global shared_counter
    for _ in range(1000):
        shared_counter += 1  # Not thread-safe

# --------- Correct functions ---------
def safe_fetch_api(endpoint):
    sock = socket.socket()
    try:
        sock.connect(('api.com', 80))
        sock.send(endpoint.encode())
        return sock.recv(1024)
    finally:
        sock.close()  # Properly closed

def process_text(data):
    if data is None:
        return ""
    return data.upper()  # Safe

def increment_safe(lock):
    global shared_counter
    for _ in range(1000):
        with lock:
            shared_counter += 1  # Thread-safe

File: GraphCodeBert/examples/TrialMixed.java
// TrialMixed.java

import java.io.FileReader;
import java.io.IOException;

public class TrialMixed {

    private int counter = 0;

    // --------- Defective methods ---------
    public String readFileUnsafe(String filename) throws IOException {
        FileReader fr = new FileReader(filename);
        // Missing fr.close()
        char[] data = new char[100];
        fr.read(data);
        return new String(data);
    }

    public void processStringUnsafe(String s) {
        if (s == null) {
            System.out.println(s.length()); // NullPointerException
        }
    }

    public void incrementUnsafe() {
        counter++; // Not synchronized
    }

    // --------- Correct methods ---------
    public String readFileSafe(String filename) throws IOException {
        FileReader fr = new FileReader(filename);
        try {
            char[] data = new char[100];
            fr.read(data);
            return new String(data);
        } finally {
            fr.close(); // Properly closed
        }
    }

    public void processStringSafe(String s) {
        if (s == null) {
            System.out.println(""); // Safe
        } else {
            System.out.println(s.length());
        }
    }

    public synchronized void incrementSafe() {
        counter++; // Thread-safe
    }
}
